Add native ESPHome API component

OttoWinter opened the pull request:
## Description:

This is a MVP for a native API for [esphomelib](https://esphomelib.com/esphomeyaml/index.html).
Actually it's even less than MVP: there's no way to configure it since this will be a config-flow only integration (only way to configure it with this PR is by editing config entries storage).

This PR includes the basic functionality for connecting to the API and a sample sensor platform for seeing how platforms would look like with the helpers.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7873

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

OttoWinter committed:
Create esphomelib component

OttoWinter committed:
Update requirements

OttoWinter committed:
Remove python 2 string literals

OttoWinter committed:
MVP

OttoWinter committed:
Remove config flow

OttoWinter committed:
Remove config flow translations

MartinHjelmare commented on .coveragerc:
```diff
@@ -117,6 +117,9 @@ omit =
     homeassistant/components/ecovacs.py
     homeassistant/components/*/ecovacs.py
 
+    homeassistant/components/esphomelib.py
```
Adjust the path to match the package path.

MartinHjelmare commented on CODEOWNERS:
```diff
@@ -184,6 +184,8 @@ homeassistant/components/*/edp_redy.py @abmantis
 homeassistant/components/edp_redy.py @abmantis
 homeassistant/components/eight_sleep.py @mezz64
 homeassistant/components/*/eight_sleep.py @mezz64
+homeassistant/components/esphomelib.py @OttoWinter
```
See above.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+DOMAIN = 'esphomelib'
+REQUIREMENTS = ['aioesphomeapi==1.0.0']
+
+# The key to store the APIClient in hass.data
```
You have named the constants very good, so we don't need the comments here.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+    return True
+
+
+async def async_setup(hass, config):
```
Please move this above `async_setup_entry`. That's where we usually have it, and it also in line with the order that functions will be called.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+    entry_data[DATA_AVAILABLE] = False
+
+    try_connect = await _setup_auto_reconnect_logic(
+        hass, cli, entry.entry_id, host)
```
If we pass the entry here instead, we could make the `get_entry_data` more strict about what argument types it should accept.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+        # indefinitely
+        await try_connect(is_disconnect=False)
+
+    hass.async_add_job(complete_setup())
```
```suggestion
    hass.async_create_task(complete_setup())
```

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+    async def complete_setup():
+        """Complete the config entry setup."""
+        for component in HA_COMPONENTS:
+            await hass.config_entries.async_forward_entry_setup(
```
Store the tasks in a list and use `asyncio.wait`, since order isn't important here.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+        if tries != 0:
+            # If not first re-try, wait and print message
+            wait_time = min(2**tries, 300)
+            _LOGGER.info("Trying to reconnect in %s seconds.", wait_time)
```
```suggestion
            _LOGGER.info("Trying to reconnect in %s seconds", wait_time)
```

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+            # So therefore all these connection warnings are logged
+            # as infos. The "unavailable" logic will still trigger so the
+            # user knows if the device is not connected.
+            _LOGGER.info("Disconnected from API.")
```
```suggestion
            _LOGGER.info("Disconnected from API")
```

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+            await cli.connect()
+            await cli.login()
+        except APIConnectionError as error:
+            _LOGGER.info("Can't connect to esphomelib API for '%s' (%s).",
```
```suggestion
            _LOGGER.info("Can't connect to esphomelib API for '%s' (%s)",
```

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+                         host, error)
+            # Schedule re-connect in event loop in order not to delay HA
+            # startup. First connect is scheduled in tracked tasks.
+            data['reconnect'] = hass.loop.create_task(
```
```suggestion
            data[DATA_RECONNECT_TASK] = hass.loop.create_task(
```

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+    await _cleanup_instance(hass, config_entry)
+
+    for component in HA_COMPONENTS:
+        await hass.config_entries.async_forward_entry_unload(
```
Store tasks in a list and use `asyncio.wait`.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+                # Create new entity
+                entity = entity_type(entry.entry_id)
+                add_entities.append(entity)
+            entity.async_info_update(info)
```
It's better to use dispatcher to call entity methods. This is perhaps technically called from within the entity platform, but we should still to use dispatcher.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+
+        # Remove old entities
+        for entity in old_entities.values():
+            hass.async_add_job(entity.async_remove())
```
Send dispatcher signal to entity here too.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+                entity = entity_type(entry.entry_id)
+                add_entities.append(entity)
+            entity.async_info_update(info)
+            new_entities[info.key] = entity
```
Don't store the entity, but store a reference to the entity, eg some unique id of the entity. We shouldn't store entities in shared containers. Entities are a platform concern.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+        if state.key not in entities:
+            return
+        entity = entities[state.key]
+        entity.async_state_update(state)
```
Send a signal to the entity with dispatcher.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+        """Schedule a state update when the device settings change."""
+        entities = get_entry_data(hass, entry)[component_key]
+        for entity in entities.values():
+            if entity.hass is not None:
```
This isn't needed.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+        entities = get_entry_data(hass, entry)[component_key]
+        for entity in entities.values():
+            if entity.hass is not None:
+                entity.async_schedule_update_ha_state()
```
Send a signal here instead. In the entity connect the state scheduling callback in `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+    def async_info_update(self, info):
+        """Update the info attribute."""
+        self.info = info
+        if self.hass is not None:
```
Remove this.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+    def async_state_update(self, state):
+        """Update the state attribute."""
+        self._state = state
+        if self.hass is not None:
```
Remote this.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+        self._state = None
+
+    @callback
+    def async_info_update(self, info):
```
These callbacks should be connected to signals in `async_added_to_hass` entity coroutine method.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,338 @@
+            self.async_schedule_update_ha_state()
+
+    @callback
+    def async_state_update(self, state):
```
See above.

OttoWinter committed:
Use dispatcher more

OttoWinter committed:
Use data classes

OttoWinter committed:
Type Hints

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,372 @@
+                                 DISPATCHER_ON_DEVICE_UPDATE.format(**kwargs),
+                                 self.async_schedule_update_ha_state)
+
+    @property
```
Add a coroutine method `async_will_remove_from_hass` here that disconnects the dispatchers. When connecting to the dispatcher signal a disconnect function is returned. Save those eg in a list as an instance attribute.

MartinHjelmare replied:
See rainmachine component for example.

MartinHjelmare commented on homeassistant/components/esphomelib/__init__.py:
```diff
@@ -0,0 +1,389 @@
+async def _cleanup_instance(hass: HomeAssistantType,
+                            entry: ConfigEntry) -> None:
+    """Cleanup the esphomelib client if it exists."""
+    data = hass.data[DOMAIN].pop(entry.entry_id)  # type: RuntimeEntryData
```
Make sure to unsubscribe any remaining component wide listeners here, eg the on home assistant stop event listener.

MartinHjelmare replied:
What will happen if we remove the config entry, but don't unsubscribe the home assistant stop event listener? Will there be a KeyError here?

OttoWinter replied:
Oh sorry, I was looking for uses of `async_dispatcher_connect` and forgot that callback

OttoWinter committed:
Cleanup on remove

OttoWinter committed:
Use helper and cleanup

OttoWinter committed:
Fix HA stop listener cleanup

OttoWinter commented:
Oh and for documentation: There's no way to configure this component at the current state (if possible, I would like it to be only configurable through config flow, as an "import" step is kinda hard/impossible here - users can use `test.local` but also `192.168.178.42`, there's no way to check if that already exists).

So is it okay to add docs once this PR is done and I add the config flow PR?

MartinHjelmare commented:
We probably want to add the config flow with tests in this PR, to make it usable.

I think the code looks very good now.

OttoWinter committed:
Add config flow

OttoWinter committed:
Fix SyntaxError for Py3.5

OttoWinter committed:
Update

MartinHjelmare commented on homeassistant/components/esphomelib/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+            return self._async_get_entry()
+
+        fields = OrderedDict()
+        fields[vol.Required('host', default=self._host or vol.UNDEFINED)] = str
```
What does `vol.UNDEFINED` do here?

OttoWinter replied:
the `vol.Required` class takes some keyword arguments including `default`. By default it is the sentinel value `vol.UNDEFINED`, so in essence this is the same as

```python
if self._host is not None:
    fields[vol.Required('host')] = str
else:
    fields[vol.Required('host', default=self._host)] = str
```

just shorter. (https://github.com/alecthomas/voluptuous/blob/master/voluptuous/schema_builder.py#L1150)

MartinHjelmare commented on homeassistant/components/esphomelib/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+    """Fetch device info from API and return any errors."""
+    from aioesphomeapi.client import APIClient, APIConnectionError
+
+    cli = APIClient(asyncio.get_event_loop(), host, port, '')
```
Shouldn't we use `hass.loop` explicitly here?

OttoWinter replied:
Well a) this method will only ever be called from the HA event loop anyway. And b) if it would be called from another event loop, `asyncio.get_event_loop()` would be the correct behavior as otherwise this could result in a deadlock.

MartinHjelmare replied:
We could pass in a loop parameter to the function, and the function still doesn't know what loop it is, but the caller knows what loop will be used.

OttoWinter committed:
Lint

MartinHjelmare commented on homeassistant/components/esphomelib/config_flow.py:
```diff
@@ -0,0 +1,129 @@
+                'host': self._host,
+                'port': self._port,
+                # The API uses protobuf, so empty string denotes absence
+                'password': self._password or '',
```
It's not possible to avoid storing the password?

OttoWinter replied:
Sure but there's not really a need for such logic IMHO. Otherwise this logic would have to sit in `async_setup_entry` like this:

```yaml
password = entry.data.get(CONF_PASSWORD, '')
```

but IMO that kind of defeats the purpose of config entries (having a well-defined internal data schema that does not need any validation).

MartinHjelmare replied:
I mean, would it be possible to avoid storing the password at all, and just use what the user enters in the form to do the authentication at that time, and store something else (token) that represents the user? I'm thinking about how hue or tradfri integrations work. I'm guessing the auth schema of esplib doesn't allow that?

balloob replied:
I think what Martin means is that the username and password are exchanged for a token that is then used to make requests. That's how most web services, like oauth2, work.

OttoWinter replied:
No the API doesn't allow that in this case and I can't really do anything against that.

There's a few reasons for that:

 - The ESPs have basically no storage, so they can't store any tokens.
 - Additionally, the password will be stored in a YAML file somewhere anyway. And if the user is using the esphomeyaml Hass.io add-on that path will be `<HA_CONFIG_DIR>/esphomeyaml/<FILENAME>`. So if someone has access to HA's storage dir then they will have the password anyway.

MartinHjelmare commented on tests/components/esphomelib/test_config_flow.py:
```diff
@@ -0,0 +1,120 @@
+    """Mock out the fetch device info method."""
+    with patch(
+            'homeassistant.components.esphomelib.config_flow.'
+            'fetch_device_info') as mock_fetch:
```
Patch the library not the config flow function.

MartinHjelmare commented on tests/components/esphomelib/test_config_flow.py:
```diff
@@ -0,0 +1,120 @@
+    """Mock out the try login method."""
+    with patch(
+            'homeassistant.components.esphomelib.config_flow.'
+            'try_login') as mock_try:
```
See above.

MartinHjelmare commented on tests/components/esphomelib/test_config_flow.py:
```diff
@@ -0,0 +1,120 @@
+
+async def test_user_connection_works(hass, mock_fetch_device_info):
+    """Test we can finish a config flow."""
+    result = await hass.config_entries.flow.async_init(
```
Look at the hue config flow tests for how to initiate the flow.

OttoWinter replied:
This is based on the MQTT config flow tests :P Ok, will do

balloob commented:
Can we rename this component to `esphome` ?

OttoWinter committed:
Re-do tests

OttoWinter committed:
:sparkles: Rename to ESPHome

OttoWinter committed:
Better error message for resolve errors

OttoWinter committed:
Fix tests when aioesphomeapi not installed

MartinHjelmare commented on tests/components/esphome/test_config_flow.py:
```diff
@@ -0,0 +1,179 @@
+    result = await flow.async_step_user(user_input=None)
+    assert result['type'] == 'form'
+
+    with patch('aioesphomeapi.APIClient') as mock_client:
```
Try to break out these patches, that are repeated in the tests with only slight changes, into one or more pytest fixtures. That will make it more DRY. See tellduslive and daikin config flow tests for examples. It's possible to overwrite pytest fixtures per test if needed.

OttoWinter committed:
Refactor mock

OttoWinter committed:
Update requirements

MartinHjelmare commented on tests/components/esphome/test_config_flow.py:
```diff
@@ -0,0 +1,159 @@
+        yield mock_client
+
+
+@pytest.fixture(autouse=True)
```
Should it be auto used?

OttoWinter replied:
Otherwise the `try:` `except APIConnectionError` clause will fail because `MagicMock` doesn't inherit from `BaseException`

MartinHjelmare reviewed (commented):
We're missing a `strings.json` file in the package. This file will be the template for the translation json files.

OttoWinter committed:
Add strings.json

OttoWinter commented:
Ah sorry about that, didn't read the guide properly 😅

I moved the `.translations/en.json` to `strings.json` and ran the `script/translations_develop` script. That script copied the file back (and changed the tellduslive and zha translations). Should I have run the script or not?

MartinHjelmare commented:
The translation upload script will be run on each dev branch build by travis, as part of the deploy step (if I understand things correctly). So we don't really need to include the translation files in PRs. But to be able to test locally properly it's nice to at least have the English translation file generated.

We probably shouldn't include files from other components in this PR.

MartinHjelmare commented on .coveragerc:
```diff
@@ -117,6 +117,9 @@ omit =
     homeassistant/components/ecovacs.py
     homeassistant/components/*/ecovacs.py
 
+    homeassistant/components/esphome/*.py
```
Don't include the config flow module in coverage calculation exclusion.

MartinHjelmare reviewed (commented):
Some logic paths of the config flow are not tested:
```
(home-assistant) m:~/Dev/home-assistant$ pytest --cov-report term-missing --cov=homeassistant.components.esphome.config_flow tests/components/esphome/test_config_flow.py 
Test session starts (platform: linux, Python 3.6.5, pytest 4.0.1, pytest-sugar 0.9.2)
rootdir: /home/martin/Dev/home-assistant, inifile: setup.cfg
plugins: requests-mock-1.5.2, timeout-1.3.3, sugar-0.9.2, rerunfailures-5.0, mock-1.10.0, cov-2.6.0, aiohttp-0.3.0
collecting ... 
 tests/components/esphome/test_config_flow.py ✓✓✓✓                                                                                                                                             100% ██████████

----------- coverage: platform linux, python 3.6.5-final-0 -----------
Name                                              Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------
homeassistant/components/esphome/config_flow.py      73      4    95%   100, 116-118


Results (0.32s):
       4 passed
```

OttoWinter committed:
:tea: 100% config flow  test coverage

MartinHjelmare commented on homeassistant/components/sensor/esphome.py:
```diff
@@ -0,0 +1,62 @@
+
+    @property
+    def _static_info(self) -> 'SensorInfo':
+        return super()._static_info
```
This doesn't look necessary.

OttoWinter replied:
For the runtime it's not necessary, yes. 

The reason it's here is because the parent `_static_info` property is type annotated as returning `EntityInfo`. So then my IDE for example highlights the line `self._static_info.unit_of_measurement` as an error, as `EntityInfo` doesn't have `unit_of_measurement`.

As you might have noticed, I like type annotations 😇

MartinHjelmare commented on homeassistant/components/sensor/esphome.py:
```diff
@@ -0,0 +1,62 @@
+
+    @property
+    def _state(self) -> Optional['SensorState']:
+        return super()._state
```
See above.

MartinHjelmare reviewed (commented):
I think we're done. Just two possible clean ups left.

OttoWinter commented:
Also really want to thank you @MartinHjelmare for all your reviews ❤️

MartinHjelmare reviewed (approved):
Awesome!
